app/scripts/controllers/network/create-network-client.ts

```diff
@@ -0,0 +1,196 @@
+): { provider: SafeEventEmitterProvider; blockTracker: PollingBlockTracker } {
+  /* eslint-disable @typescript-eslint/no-require-imports,@typescript-eslint/no-shadow */
+  const fetch = global.fetch || require('node-fetch');
+  const btoa = global.btoa || require('btoa');
+
+  const rpcApiMiddleware =
+    networkConfig.type === NetworkClientType.Infura
+      ? createInfuraMiddleware({
+          network: networkConfig.network,
+          projectId: networkConfig.infuraProjectId,
+          maxAttempts: 5,
+          source: 'metamask',
+        })
+      : createFetchMiddleware({ btoa, fetch, rpcUrl: networkConfig.rpcUrl });
+
+  const rpcProvider = providerFromMiddleware(rpcApiMiddleware);
+
+  const blockTrackerOpts =
+    process.env.IN_TEST && networkConfig.type === 'custom'
+      ? { pollingInterval: SECOND }
+      : {};
+  const blockTracker = new PollingBlockTracker({
+    ...blockTrackerOpts,
+    provider: rpcProvider,
+  });
+
+  const networkMiddleware =
+    networkConfig.type === NetworkClientType.Infura
+      ? createInfuraNetworkMiddleware({
+          blockTracker,
+          network: networkConfig.network,
+          rpcProvider,
+          rpcApiMiddleware,
+        })
+      : createCustomNetworkMiddleware({
+          blockTracker,
+          chainId: networkConfig.chainId,
+          rpcApiMiddleware,
+        });
+
+  const engine = new JsonRpcEngine();
+
+  engine.push(networkMiddleware);
+
+  const provider = providerFromEngine(engine);
+
+  return { provider, blockTracker };
+}
+
+function createInfuraNetworkMiddleware({
+  blockTracker,
+  network,
+  rpcProvider,
+  rpcApiMiddleware,
+}: {
+  blockTracker: PollingBlockTracker;
+  network: BuiltInInfuraNetwork;
+  rpcProvider: SafeEventEmitterProvider;
+  rpcApiMiddleware: JsonRpcMiddleware<unknown, unknown>;
+}) {
+  return mergeMiddleware([
+    createNetworkAndChainIdMiddleware({ network }),
+    createBlockCacheMiddleware({ blockTracker }),
+    createInflightCacheMiddleware(),
+    createBlockRefMiddleware({ blockTracker, provider: rpcProvider }),
+    createRetryOnEmptyMiddleware({ blockTracker, provider: rpcProvider }),
+    createBlockTrackerInspectorMiddleware({ blockTracker }),
+    rpcApiMiddleware,
+  ]);
+}
+
+function createNetworkAndChainIdMiddleware({
+  network,
+}: {
+  network: BuiltInInfuraNetwork;
+}) {
+  if (!BUILT_IN_INFURA_NETWORKS[network]) {
+    throw new Error(`createInfuraClient - unknown network "${network}"`);
+  }
+
+  const { chainId, networkId } = BUILT_IN_INFURA_NETWORKS[network];
+
+  return createScaffoldMiddleware({
+    eth_chainId: chainId,
+    net_version: networkId,
+  });
+}
+
+const createChainIdMiddleware: JsonRpcMiddleware<unknown, unknown> = (
+  chainId: string,
+) => {
+  return (
+    req: any,
+    res: any,
+    next: () => Promise<void>,
+    end: () => Promise<void>,
+  ) => {
+    if (req.method === 'eth_chainId') {
+      res.result = chainId;
+      return end();
+    }
+    return next();
+  };
+};
```

Comment:
The `createChainIdMiddleware` function doesn't represent the middleware, it returns the middleware. So you should be able to say:

``` suggestion
const createChainIdMiddleware = (chainId: string) => {
  const chainIdMiddleware: JsonRpcMiddleware<unknown, unknown> (req, res, next, end) => {
    if (req.method === 'eth_chainId') {
      res.result = chainId;
      return end();
    }
    return next();
  };
  return chainIdMiddleware;
};
```

I guess now that I've looked at your example you can also move the type this way:

``` suggestion
const createChainIdMiddleware = (chainId: string): JsonRpcMiddleware<unknown, unknown> => {
  return (req, res, next, end) => {
    if (req.method === 'eth_chainId') {
      res.result = chainId;
      return end();
    }
    return next();
  };
};
```

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/45e028e815c9fbb7180d10e902dd130255afe740